### ISSUE TITLE
Globbing does not work

### ISSUE THREAD

**Issue by PulsarFox** [solid/node-solid-server#697]
When I try to access several resource with globbing (like /profile/*) the request gets stuck then fails 3min after with a status 0.

I've tried to fix that but I've only managed to pinpoint where it fails (there is no error message, just a stuck request)
It's in the Allow function in allow.js (used in https://github.com/solid/node-solid-server/blob/master/lib/handlers/get.js#L203)

The thing is if I call a log debug message between the 9th and the 10th line in https://github.com/solid/node-solid-server/blob/master/lib/handlers/allow.js on a /profile/* request (containing a card and card.acl, so 2 files), I get this debug message two times, so nothing wrong here.

But on the allowHandler function, only one debug message is displayed, even if the Allow function is called two times... 

I'm using an express server and here is my config options : 
```
    cache: 0, 
    serverUri:"https://localhost:"+port,
    live: true, 
    root: './LDP/', 
    configPath:'./LDP/.config',
    dbPath:'./LDP/.db',
    secret: 'node-ldp', 
    sslCert: '../localhost.cert', 
    sslKey: '../localhost.key', 
    mount: '/', 
    auth:'oidc',
    multiuser:true,
    webid: true, 
    suffixAcl: '.acl',
    corsProxy: '/proxy', 
    errorHandler: function(err, req, res, next){
        console.log(err)
    },
    errorPages: false
```

**Comment by dmitrizagidulin** [comment]
Thanks @PulsarFox! We will investigate.

**Comment by RubenVerborgh** [comment]
To what extent do we support globbing? Is this documented?
What current applications use globbing?

**Comment by melvincarvalho** [comment]
Globbing has been 100% part of the solid spec for years.  It provides an important optimization for client side apps.

https://github.com/solid/solid-spec/blob/master/api-rest.md#globbing-inlining-on-get

It's supported by gold, and should be part of the test coverage in node solid server.

**Comment by RubenVerborgh** [comment]
> Globbing has been 100% part of the solid spec for years.

Okay, thanks for the pointer!

> It provides an important optimization for client side apps.

Which apps are currently using it?

Note that it's an optimization, so not a necessity. This might affect prioritization.

> should be part of the test coverage in node solid server.

That's where I'm not sure: I'm seeing lots of globbing code being commented out, and some globbing tests being skipped. I don't know whether globbing was ever fully implemented on node-solid-server, and if it is not, we should wonder how high on the priority list it should be.

**Comment by melvincarvalho** [comment]
It's fully implemented in gold.  Globbing is core to solid.  NSS cannot be called a solid compliant server if globbing doesnt work.  Commenting out tests was possibly in order to pass travis CI.

I have written couple of apps that use it, one chat, one social.  There could be more.  100s of async get requests coming back in a random order is less practical than a glob.

As a background, essentially I was forced to put app writing on hold while bugs to globbing, websockets and auth have been outstanding.  About 2 years ago everything was working in gold, except for some patchy websocket support.  I patched the web socket and ran my own server, but this was the only working solid server.  However, solid is not much fun if only one server is compliant for apps.  After that OIDC was added, 

So the idea, is that node solid server should be compatible with gold in TLS mode.  If OIDC is added to gold, then it will be compatible in both modes.  This will provide a uniform experience for app developers across servers.

**Comment by RubenVerborgh** [comment]
> It's fully implemented in gold.

Yes, but we're not aiming to replicate gold.

> Globbing is core to solid.

Citation needed. That's exactly the question I was asking :-)

> NSS cannot be called a solid compliant server if globbing doesnt work.

But maybe we need to reconsider the spec.
We should have a discussion on spec level.

> Commenting out tests was possibly in order to pass travis CI.

It was.

> I have written couple of apps that use it, one chat, one social.

Do you have URLs? Might give us an idea of how globbing is used.

> 100s of async get requests coming back in a random order is less practical than a glob.

"Practical" can just be fixed with a client-side library.
(Playing devil's advocate here.)

> As a background, essentially I was forced to put app writing on hold while bugs to globbing, websockets and auth have been outstanding.

But other people were not.

I hear you loud and clear, globbing is important to you, but I'm trying to gauge how important it is for the ecosystem, given that it is an optimization.

> So the idea, is that node solid server should be compatible with gold in TLS mode.

I'd say both have to be compatible with the spec; and maybe we need to revisit the spec if certain features aren't crucial. Or at least consider layered compatibility, where optimizations can be carried out by either the client or the server.

**Comment by melvincarvalho** [comment]
What can I say, it's part of the core solid spec which gathered consensus over many years and 100s of meetings.  It's tough to pack all that into github issue.

I wrote two apps using globbing, there could be more out there, a lot of work went into them, but bugs in the server, and breaking changes to angular prevented an optimal user experience.  That said, as a proof of concept they worked quite well.

Chat app : [source](https://github.com/melvincarvalho/webid.im) | [Screenshot](https://raw.githubusercontent.com/melvincarvalho/webid.im/master/paper/paper.png)

Social app : [source](https://github.com/solid-social/timeline) | [Screenshots](https://melvincarvalho.gitbooks.io/solid-social/content/appendixa.html)

There may be more apps that use it, or even the data browser.  In terms of prioritization, we have at various times been close to having a bug free server (modulo the new OIDC work).  I think the priority should be fixing all the bugs as part of stabilization.

There is the [test](https://github.com/linkeddata/gold/blob/2c39f9024bcdc30726ea102bfd75a13569f63500/server_test.go#L1177) in gold if that helps understand a bit better.

**Comment by RubenVerborgh** [comment]
Nice stuff, we should have a look at your apps (sent an IM about that).

I would suggest we're not too dogmatic about decisions; insights evolve and situations do as well. For instance, querying might be a replacement for globbing needs.

At the moment, we have evidence of 2 applications using globbing. What we still need is an understanding of how crucial globbing is for these apps (since it is an optimization, we know that the app can work without it, but might require some adaptions), how impactful the apps are on the ecosystem, and what other apps might benefit of globbing.

So instead of having a dogmatic “we need globbing”, what we should examine is “what optimization needs to apps have”, and how can these be met best. Given that querying is on the roadmap anyways, it might (or might not) be the better answer altogether.

**Comment by melvincarvalho** [comment]
Implementing a spec is not being overly dogmatic, IMHO.  The dogma lies in the typical process of consensus leading to a spec, and having test coverage.  Commenting out tests in order to pass Travis gives the illusion of a passing server, when it is not.  One could reasonably argue, why have tests at all.

At a higher level, IMHO, we benefit most from fixing the bugs in the server, something we have been close to for some time, but not quite over the line.  I appreciate bugs can be very hard to track down, let alone fix.  That basically includes auth issues, globbing, delegation and on-boarding.

By all means specs are there to be challenged, but that is an orthogonal process, which is related to coming to a consensus over time.  The default position should be to implement the spec.

**Comment by RubenVerborgh** [comment]
> Commenting out tests in order to pass Travis gives the illusion of a passing server

We're not commenting out, we are marking them as skipped. The logs show passed/failing/skipped. The server passes everything that is implemented; skipped tests point to non-implemented features.

> One could reasonably argue, why have tests at all.

We have tests to detect whether a commit doesn't break something that worked before.

Skipped test specify behavior that you might want to implement at some point. Having builds fail systematically because of something that is not implemented would mask a commit actually breaking a build.

>  I appreciate bugs can be very hard to track down, let alone fix.

Yeah, but the question here is more: was globbing ever fully implemented? I think the answer is no. So then it's logical that some things don't work, and this is not a bug but a "non-implemented".

So perhaps there should be an issue of "implement globbing", which we might or might not assign to a future node-solid-server version.

**Comment by melvincarvalho** [comment]
> the question here is more: was globbing ever fully implemented?

Yes, it was.

**Comment by RubenVerborgh** [comment]
> Yes, it was.

That contradicts @dmitrizagidulin's and my knowledge about node-solid-server.
Could you point me to a version number or commit that has complete globbing implementation? Then we'll take it from there.

**Comment by melvincarvalho** [comment]
A few points here and to summarize, in no particular order : 

- Globbing was implemented some years ago, before you joined the project, possibly even before dmitri (have a look at git blame for more details)

- Globbing is part of the core solid spec

- Globbing is not only a useful optimization, it is a first class citizen as a URI.  Consider a named graph from the glob URI.  That would be teh 

- Globbing is implemented in more than one app, which are blocked by server issues, such as this one

- Globbing is implemented in other servers, such as gold, possibly even ldphp, so for a consistent long term developer experience, all servers should implement the core spec

- The core spec is evolving, but we have had no substantive proposals around alternatives to globbing, that could happen in parallel tho.

I dont imagine it's a huge amount of code, and it was working before.  We want node solid server to be a compliant and work with other implementations, perhaps leading to solid becoming a w3c rec, and other nice things! :)

**Comment by RubenVerborgh** [comment]
Turns out that fixing globbing was easier and more productive than arguing with @melvincarvalho 😉

**Comment by melvincarvalho** [comment]
Thanks @RubenVerborgh! 

Was just trying to give the background information, which you requested.

**Comment by RubenVerborgh** [comment]
Of course. Glob along, and let's have a look at porting your apps to solid-auth-client, they make nice demos for the whole community.
